### DISCUSSION TITLE
Showing macro recording was started and finished

### DISCUSSION THREAD

**Discussion by yart** [folke/noice.nvim#795] [Q&A] [answered]
### Did you check docs and existing issues?

- [X] I have read all the noice.nvim docs
- [X] I have searched the existing issues of noice.nvim
- [X] I have searched the existing issues of plugins related to this issue

### Neovim version (nvim -v)

NVIM v0.9.5 Build type: Release LuaJIT 2.1.1713773202

### Operating system/version

ArchLinux Kernel version: 6.8.9-arch1-1

### Describe the bug

I'm using NvChad 2.0.

<details><summary>Here is my settings for noice and lualine</summary>

```lua
return {
  {
    "folke/noice.nvim",
    event = "VeryLazy",
    opts = {
      -- add any options here
    },
    dependencies = {
      -- if you lazy-load any plugin below, make sure to add proper `module="..."` entries
      "MunifTanjim/nui.nvim",
      -- OPTIONAL:
      --   `nvim-notify` is only needed, if you want to use the notification view.
      --   If not available, we use `mini` as the fallback
      "rcarriga/nvim-notify",
    },
    config = function()
      require("noice").setup({
        lsp = {
          -- override markdown rendering so that **cmp** and other plugins use **Treesitter**
          override = {
            ["vim.lsp.util.convert_input_to_markdown_lines"] = true,
            ["vim.lsp.util.stylize_markdown"] = true,
            ["cmp.entry.get_documentation"] = true, -- requires hrsh7th/nvim-cmp
          },
        },
        -- you can enable a preset for easier configuration
        presets = {
          bottom_search = false, -- use a classic bottom cmdline for search
          command_palette = true, -- position the cmdline and popupmenu together
          long_message_to_split = true, -- long messages will be sent to a split
          inc_rename = false, -- enables an input dialog for inc-rename.nvim
          lsp_doc_border = false, -- add a border to hover docs and signature help
        },
        cmdline = {
          view = "cmdline",
        },
        -- routes = {
        --   view = "mini",
        --   filter = {
        --     event = "msg_showmode",
        --     any = {
        --       { find = "recording" },
        --     },
        --   },
        -- },
      })
    end,
    keys = {
      { "<leader>Nt", "<cmd>Noice telescope<cr>", desc = "Show all notifications using Telescope" },
      { "<leader>Nl", "<cmd>Noice last<cr>", desc = "Show the last notification" },
    }
  },
  {
    'nvim-lualine/lualine.nvim',
    dependencies = {
      'nvim-tree/nvim-web-devicons',
      "folke/noice.nvim",
    },
    config = function()
      require("lualine").setup({
        sections = {
          lualine_x = {
            {
              require("noice").api.statusline.mode.get,
              cond = require("noice").api.statusline.mode.has,
              color = { fg = "#ff9e64" },
            }
          },
        },
      })
    end
  },
}

```
</details>

When I type `q` in normal mode to start recording a macro, I don't see any notification or another sign, that it was started. But the recording goes as usual. And I can use the recorded macros as usual.

Anyway, it's slightly annoying when I can't control this process. Of course, in many cases I don't need any visualization of this process, but sometimes it's needed.

### Steps To Reproduce

1. Install NvChad following the instruction [here](https://nvchad.com/docs/quickstart/install/)
2. Load my config to ~/.config/nvim/lua/custom/plugins.lua
3. Try to start recording a macro typing `q`
4. Just be sure, that you don't have any notification about the macro recording starting

### Expected Behavior

The macro recording notification must be shown.

### Repro

```Lua
-- DO NOT change the paths and don't remove the colorscheme
local root = vim.fn.fnamemodify("./.repro", ":p")

-- set stdpaths to use .repro
for _, name in ipairs({ "config", "data", "state", "cache" }) do
  vim.env[("XDG_%s_HOME"):format(name:upper())] = root .. "/" .. name
end

-- bootstrap lazy
local lazypath = root .. "/plugins/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", lazypath, })
end
vim.opt.runtimepath:prepend(lazypath)

-- install plugins
local plugins = {
  "folke/tokyonight.nvim",
  "folke/noice.nvim",
  -- add any other plugins here
}
require("lazy").setup(plugins, {
  root = root .. "/plugins",
})

vim.cmd.colorscheme("tokyonight")
-- add anything else here
```

**Comment by ilias777** [comment]
Try this:

```lua
routes = {
    {
        view = 'notify',
        filter = {
            event = 'msg_showmode',
            any = {
                { find = 'recording' },
            },
        },
    },
}
```

This is described [here](https://github.com/folke/noice.nvim/wiki/Configuration-Recipes#show-recording-messages)

**Reply by bbil** [reply]
This works ok, and is much better than the recipe from the link. The recipe from the link will also create notifications for entering insert mode!

But I am also not having any luck with integrating this into Lualine like that link suggests... Get an error message around `noice` not being found.

**Reply by dandyrow** [reply]
I've also found the same issue when trying to use with Lualine. My LSP is saying that the statusline property of noice is deprecated.

**Reply by dandyrow** [reply]
I've managed to get lualine functionality working. I think it was because my options weren't in a function but rather just a table in the lualine config file. I have converted to a lua function and it now works.

**Comment by yart** [comment]
Guys, I was incorrect when have been writing this issue. I see the message like `@a recording`, but I don't see any message after the `q` key is pressed. So, I can press `q` randomly and start recording a macro. On the one side this is not a problem: I can stop it any moment, but on the other side this is annoying, because it looks like I'm losing the control.

I've tried to catch the moment when I pressed `q` and show a notification with this, but it looks like pressing `q` pausing other processes and I see both notifications (about `pressed` and the macro recording) at the same time, when chosen the register.

So, if somebody knows a solution of how to indicate this state, I would be really grateful to this person.

**Reply by ilias777** [reply]
I don't understand what you mean.

In normal Neovim, without noice.nvim and nvim-notify, if you press `q`, only one time a message appears `recordong @{key}` at the bottom and that is. After a keysequence and pressing `q` again the message disappears and you have recorded the macro. There is no other message that you are done with the macro.

Same with noice.nvim and nvim-notify. With the code snippet above, the message appears only once as notify message.

I don't know the integration with lualine. I use feline as my statusline and it shows the recording message there.

Take a look:

https://github.com/folke/noice.nvim/assets/39212564/d7c08e3f-e7e1-4907-9f7f-add7e8f9750b

I press here `qa` and the message appears as long I don't press `q`. If I'm done with my macro, I press `q` again and the message disappears. So I know the recording is finished. For me is this the best way to know that I am writing a macro.

**Comment by yart** [comment]
@ilias777 let me explain: when I press `q` in nvim w/o any plugins, it firstly shows the `q` letter on the right side of the command line. When I press another letter (to put a register), it removes the `q` letter from the right side of the command line and shows `@<key> recording` on the left side of the command line instead.

If I misclicked the `q` key, I can see this letter `q` on the right side of the command line, and it indicates me, that nvim (or vim) is ready to record the macro. I can cancel it with ECS or continue to create a macro because I see what is happening right now. 

But in case with hidden command line, I can't see this indicator on the right side of the command line, and sometimes it's annoying me.

So this is what I mean.

**Reply by ilias777** [reply]
Oh, I see. Ok what about this &rarr; Send the commands to the mini view like this:

```lua
routes = {
  {
    view = 'mini',
    filter = {
      event = 'msg_showcmd'
    }
  }
}
```

Every time you press `q` it shows at the bottom right corner in the mini view. With this you can see also the commands like `ciw`, `yiw` and so on, same as Neovim without any plugins.

If you want to see only the `q`, you can use this:

```lua
routes = {
  {
    view = 'mini',
    filter = {
      event = 'msg_showcmd'
      any = {
        { find = 'q$' } -- $ to indicate that q is the last letter
      }
    }
  }
}
```

Try it out. Maybe this is what you want.

**Reply by yart** [reply]
@ilias777 it almost works. I did the following:

```lua
        routes = {
          {
            view = "notify",
            filter = { event = "msg_showmode" },
          },
          {
            view = 'mini',
            filter = {
              event = 'msg_showcmd',
              any = {
                { find = '[mq]$' }
              },
            },
          },
```

But when I press one of these keys the first time, it doesn't show anything, while I put a register. After that, it shows me both letters. In case with `q`, it also shows me the letter when I finish a macro recording.

As I told, it looks like nvim pauses any process while waiting for a register letter input. Then it continues work as usual.

Another question: how can I combine events? I want to show notification when I'm stopping a macro recording. Sorry if the question is stupid, but this day was a bit long for me and my brain is tired.

**Reply by ilias777** [reply, marked as answer]
> But when I press one of these keys the first time, it doesn't show anything, while I put a register. After that, it shows me both letters. In case with q, it also shows me the letter when I finish a macro recording.
>
> As I told, it looks like nvim pauses any process while waiting for a register letter input. Then it continues work as usual.

Yes, this is normal behavior. If you press `q`, Neovim is waiting to press the next letter **and then** it shows the message `recording @{key}`.

> Another question: how can I combine events? I want to show notification when I'm stopping a macro recording. Sorry if the question is stupid, but this day was a bit long for me and my brain is tired.

No, your question is not stupid. To show a message if you stop a macro, you can use a `autocmd`, like this:

```lua
local macro_group = vim.api.nvim_create_augroup('MacroRecording', { clear = true })
vim.api.nvim_create_autocmd('RecordingLeave', {
    group = macro_group,
    callback = function()
        -- Display a message when macro recording stops
        print('Macro recording stopped')
    end,
})
```

https://github.com/folke/noice.nvim/assets/39212564/0a729815-4194-4b46-8d3f-d77272625efc

Keep in mind that this auto command don't trigger a third time the `RecordingLeave` event and I don't know why!

I think that is, right?

**Reply by yart** [reply]
@ilias777 sorry for the long answer — I was busy. And thank you for the idea, that is!

And I have a similar question, but about numbers for motions in normal mode. When I'm typing, for example, `11j` or `317g` it doesn't show me the numbers until I press a movement key. Have you an idea how to show the numbers before I pressed any other key?

**Reply by ilias777** [reply]
To see the numbers before the movement key, it's not possible if you send the `msg_showcmd` event to the mini view.

I tried LazyVim *(because LazyVim uses `lualine.nvim` and I use `feline.nvim`)* to see how folke has setting up Lualine to display the `showcmd` and Lualine in this distro has same functionality like `msg_showcmd` event in the mini view by noice.

The only way to see the numbers first, is when you have `cmdheight` set to `1`. But with noice you can't use `cmdheight`, because noice automatically set this option to `0`. Only if you disable noice you can set `cmdheight` to `1` and see the numbers first like `11j` or `11k`.

You can see folkes comment &rarr; [here](https://github.com/folke/noice.nvim/issues/305#issuecomment-1373979521)

**Reply by yart** [reply]
This is really sad info… Hmm… Looks like we need to commit something into the nvim repo to bypass zero `cmdheight` and achieve a result. Ok. Will learn it.

Anyway, I thank you for the information you gave me. It reduces my time to search the info.

**Reply by urizennnn** [reply]
> > But when I press one of these keys the first time, it doesn't show anything, while I put a register. After that, it shows me both letters. In case with q, it also shows me the letter when I finish a macro recording.
> > As I told, it looks like nvim pauses any process while waiting for a register letter input. Then it continues work as usual.
> 
> Yes, this is normal behavior. If you press `q`, Neovim is waiting to press the next letter **and then** it shows the message `recording @{key}`.
> 
> > Another question: how can I combine events? I want to show notification when I'm stopping a macro recording. Sorry if the question is stupid, but this day was a bit long for me and my brain is tired.
> 
> No, your question is not stupid. To show a message if you stop a macro, you can use a `autocmd`, like this:
> 
> ```lua
> local macro_group = vim.api.nvim_create_augroup('MacroRecording', { clear = true })
> vim.api.nvim_create_autocmd('RecordingLeave', {
>     group = macro_group,
>     callback = function()
>         -- Display a message when macro recording stops
>         print('Macro recording stopped')
>     end,
> })
> ```
> 
>  stop-recording-message.mov 
> Keep in mind that this auto command don't trigger a third time the `RecordingLeave` event and I don't know why!
> 
> I think that is, right?

for some reason mine doesn't show ```recording @key``` and  the macro recorded persists amongst all nvim sessions

**Reply by ilias777** [reply]
See `:h shortmess`. May that is the problem. Type `:set shortmess?` and see the output. If `q` is in the string, then you don't see any recording message.

From the `:h shortmess` help pages:

> q	do not show "recording \@ a" when recording a macro

**Reply by urizennnn** [reply]
it's not there

**Reply by ilias777** [reply]
Do you have this in your noice configuration?

```lua
routes = {
    {
        view = 'mini',
        filter = {
            event = 'msg_showmode',
            any = {
                { find = 'recording' },
            },
        },
    },
}
```

Screenshot:

<img width="1315" alt="recording" src="https://github.com/folke/noice.nvim/assets/39212564/52f34f72-cd75-44cb-89dc-30e63635191e">

If the message still not showing up, then you have to see which plugin causes the problem.

**Reply by urizennnn** [reply]
nah i got this
```lua
	routes = {
		{
			view = "notify",
			filter = { event = "msg_showmode" },
		},
		{
			view = "mini",
			filter = {
				event = "msg_showcmd",
				any = {
					{ find = "[mq]$" },
				},
			},
		}, -- see section on routes
	}```
	but i changed it to that now and still doesn't work 
but if i deactivate noice it works

**Reply by ilias777** [reply]
I have also noice installed. With your code I have this:

<img width="1462" alt="screenshot" src="https://github.com/folke/noice.nvim/assets/39212564/2271acab-009a-493e-967f-c05d3c1ee9d7">

Make sure you have `nvim-notify` as a dependency installed in the noice configuration or regular installed in your neovim config:

```lua
{
  "folke/noice.nvim",
  event = "VeryLazy",
  opts = {
    -- options
  },
  dependencies = {
    "MunifTanjim/nui.nvim",
    "rcarriga/nvim-notify",
    }
}
```

Then make sure the notify option in the noice opts are enabled.

```lua
opts = {
    notify = {
        enabled = true,
        view = "notify",
    },
},
```

**Reply by ilias777** [reply]
@urizennnn I tried your neovim config. You don't have the routes option set in your noice configuration.

This is your neovim config with the code added:

<img width="1462" alt="urizennn-config" src="https://github.com/folke/noice.nvim/assets/39212564/ec18f227-3b59-4587-8c5b-9d81d0de1ddf">

Put this code in your noice config and you are done! 😉👍🏻

**Comment by urizennnn** [comment]
this also works it's just that recording thing that doesn't show up
```lua 
local macro_group = vim.api.nvim_create_augroup("MacroRecording", { clear = true })
vim.api.nvim_create_autocmd("RecordingLeave", {
	group = macro_group,
	callback = function()
		-- Display a message when macro recording stops
		print("Macro recording stopped")
	end,
})
```
This is so weird i have all that on my local machine I haven't pushed to remote yet everything pops up nicely except the recording 
![image](https://github.com/folke/noice.nvim/assets/131607385/8a858c27-98f9-4808-bdb5-a21d4bf9bc87)

**Reply by ilias777** [reply]
I tested with your neovim config again:

https://github.com/folke/noice.nvim/assets/39212564/846dec2d-a2bf-45e9-ac24-caf6f055b5d9

It works with the stoppen macro message. The message show up.

**Reply by urizennnn** [reply]
yes the stop thing works its the recording that's not working 
but if i disable noice it works 
i don't know if lualine has some business in it or some setup in noice is wrong

**Reply by ilias777** [reply]
I already answered you. Look &rarr; [here](https://github.com/folke/noice.nvim/discussions/795#discussioncomment-9899472). Maybe you missed this.

Put this code in your noice configuration:

```lua
opts = {
    routes = {
        {
            view = 'mini',
            filter = {
                event = 'msg_showmode',
                any = {
                    { find = 'recording' },
                },
            },
        },
    },
}
```

and then it works.

**Reply by urizennnn** [reply]
It is there

**Reply by urizennnn** [reply]
```lua
require("noice").setup({
	cmdline = {
		enabled = true, -- enables the Noice cmdline UI
		view = "cmdline_popup", -- view for rendering the cmdline. Change to `cmdline` to get a classic cmdline at the bottom
		opts = {
			routes = {
				{
					view = "mini",
					filter = {
						event = "msg_showmode",
						any = {
							{ find = "recording" },
						},
					},
				},
			},
		}, -- global options for the cmdline. See section on views
		format = {
			-- conceal: (default=true) This will hide the text in the cmdline that matches the pattern.
			-- view: (default is cmdline view)
			-- opts: any options passed to the view
			-- icon_hl_group: optional hl_group for the icon
			-- title: set to anything or empty string to hide
			cmdline = { pattern = "^:", icon = "", lang = "vim" },
			search_down = { kind = "search", pattern = "^/", icon = " ", lang = "regex" },
			search_up = { kind = "search", pattern = "^%?", icon = " ", lang = "regex" },
			filter = { pattern = "^:%s*!", icon = "$", lang = "bash" },
			lua = { pattern = { "^:%s*lua%s+", "^:%s*lua%s*=%s*", "^:%s*=%s*" }, icon = "", lang = "lua" },
			help = { pattern = "^:%s*he?l?p?%s+", icon = "" },
			input = {}, -- Used by input()
			-- lua = false, -- to disable a format, set to `false`
		},
	},
	messages = {
		enabled = true, -- enables the Noice messages UI
		view = "notify", -- default view for messages
		view_error = "notify", -- view for errors
		view_warn = "notify", -- view for warnings
		view_history = "messages", -- view for :messages
		view_search = "virtualtext", -- view for search count messages. Set to `false` to disable
	},
	popupmenu = {
		enabled = true, -- enables the Noice popupmenu UI
		backend = "nui", -- backend to use to show regular cmdline completions
		kind_icons = {}, -- set to `false` to disable icons
	},
	redirect = {
		view = "popup",
		filter = { event = "msg_show" },
	},
	commands = {
		history = {
			view = "split",
			opts = { enter = true, format = "details" },
			filter = {
				any = {
					{ event = "notify" },
					{ error = true },
					{ warning = true },
					{ event = "msg_show", kind = { "" } },
					{ event = "lsp", kind = "message" },
				},
			},
		},
		last = {
			view = "popup",
			opts = { enter = true, format = "details" },
			filter = {
				any = {
					{ event = "notify" },
					{ error = true },
					{ warning = true },
					{ event = "msg_show", kind = { "" } },
					{ event = "lsp", kind = "message" },
				},
			},
			filter_opts = { count = 1 },
		},
		errors = {
			view = "popup",
			opts = { enter = true, format = "details" },
			filter = { error = true },
			filter_opts = { reverse = true },
		},
		all = {
			view = "split",
			opts = { enter = true, format = "details" },
			filter = {},
		},
	},
	notify = {
		enabled = true,
		view = "notify",
	},
	lsp = {
		progress = {
			enabled = true,
			format = "lsp_progress",
			format_done = "lsp_progress_done",
			throttle = 1000 / 30, -- frequency to update lsp progress message
			view = "mini",
		},
		override = {
			["vim.lsp.util.convert_input_to_markdown_lines"] = false,
			["vim.lsp.util.stylize_markdown"] = false,
			["cmp.entry.get_documentation"] = false,
		},
		hover = {
			enabled = true,
			silent = false,
			view = nil,
			opts = {}, -- merged with defaults from documentation
		},
		signature = {
			enabled = true,
			auto_open = {
				enabled = true,
				trigger = true, -- Automatically show signature help when typing a trigger character from the LSP
				luasnip = true, -- Will open signature help when jumping to Luasnip insert nodes
				throttle = 50, -- Debounce lsp signature help request by 50ms
			},
			view = nil,
			opts = {}, -- merged with defaults from documentation
		},
		message = {
			enabled = true,
			view = "notify",
			opts = {},
		},
		documentation = {
			view = "hover",
			opts = {
				lang = "markdown",
				replace = true,
				render = "plain",
				format = { "{message}" },
				win_options = { concealcursor = "n", conceallevel = 3 },
			},
		},
	},
	markdown = {
		hover = {
			["|(%S-)|"] = vim.cmd.help, -- vim help links
			["%[.-%]%((%S-)%)"] = require("noice.util").open, -- markdown links
		},
		highlights = {
			["|%S-|"] = "@text.reference",
			["@%S+"] = "@parameter",
			["^%s*(Parameters:)"] = "@text.title",
			["^%s*(Return:)"] = "@text.title",
			["^%s*(See also:)"] = "@text.title",
			["{%S-}"] = "@parameter",
		},
	},
	health = {
		checker = true, -- Disable if you don't want health checks to run
	},
	presets = {
		bottom_search = false, -- use a classic bottom cmdline for search
		command_palette = false, -- position the cmdline and popupmenu together
		long_message_to_split = false, -- long messages will be sent to a split
		inc_rename = false, -- enables an input dialog for inc-rename.nvim
		lsp_doc_border = false, -- add a border to hover docs and signature help
	},
	throttle = 1000 / 30, -- how frequently does Noice need to check for UI updates? This has no effect when in blocking mode.
	views = {}, -- see section on views
	routes = {
		{
			view = "mini",
			filter = {
				event = "msg_showmode",
				any = {
					{ find = "recording" },
				},
			},
		},
	},
	status = {}, -- see section on statusline components
	format = {}, -- see section on formatting
})```

**Reply by ilias777** [reply]
As I said, I used **your** nvim configuration in my pc with this code added the noice options and it work. You can see that in the screenshot that I posted yesterday. 

Delete the `opts table` that begins at line `5` in your reply. You have already routes defined at the bottom of your code.

**Reply by urizennnn** [reply]
well imma just look for the error for that or whichever plugin is disrupting that

**Comment by V4G4X** [comment]
I have chosen to use Snacks.nvim's notifier.

That has:
```lua
notifier = { enabled = true }
```
In the `snacks.lua` config, and:
```lua
notify = { enabled = false }
```
In the `noice.lua` config.

How am I supposed to disable notifications from Noice overall but still get a notification for my macro usage?

**Comment by ErlanRG** [comment]
I like how the `search_down` and `search_up` look like in the cmdline, so I did something like this to mimic that with the record macro message:

```lua
config = function()
    local colors = require 'catppuccin.palettes.mocha'
    vim.api.nvim_set_hl(0, 'MacroRecordingIcon', { fg = colors.red, bold = true })

    require('noice').setup {
        routes = {
            {
                view = 'cmdline',
                filter = { event = 'msg_showmode', any = { { find = 'recording' } } },
                opts = { format = { { '󰑊', hl_group = 'MacroRecordingIcon' }, ' ', '{message}' } },
            },
        },
    }
end
```

This not only sends the recording notification to its default location, but add a small icon with some nice colors.


<img width="142" height="40" alt="image" src="https://github.com/user-attachments/assets/5f72e6d4-0866-41b6-bec6-07edaa6fbc37" />


Another reason why is I feel like the statusline approach using noice's api feels a laggy. It takes some time for the recording message to appear and go away. This way, it feels more like the native behavior
